include/CL/sycl/buffer/detail/buffer.hpp

```diff
@@ -219,18 +318,64 @@ class buffer : public detail::buffer_base,
   }
 
 
-  /** Set the weak pointer to copy back data on buffer deletion
+  /** Set the weak pointer as destination for write-back on buffer destruction.
+  */
+  void set_final_data(std::weak_ptr<T> && final_data) {
+    shared_pointer = boost::none;
+    final_write_back = [=] {
+      if (auto sptr = final_data.lock()) {
+        std::copy_n(access.data(), access.num_elements(), sptr.get());
+      }
+    };
+  }
+
+
+  /** Provide destination for write-back on buffer destruction as a
+   *  shared pointer.
+   */
+  void set_final_data(std::shared_ptr<T> && final_data) {
+    shared_pointer = final_data;
+    final_write_back = [=] {
+      std::copy_n(access.data(), access.num_elements(), (*shared_pointer).get());
+    };
+  }
+
 
-      \todo Add a write kernel dependency on the buffer so the buffer
-      destructor has to wait for the kernel execution if the buffer is
-      also accessed through a write accessor
+  /** Provide destination for write-back on buffer destruction as a pointer.
   */
-  void set_final_data(weak_ptr_class<T> && finalData) {
-    final_data = finalData;
+  void set_final_data(T* final_data) {
+    shared_pointer = boost::none;
+    final_write_back = [=] {
+      std::copy_n(access.data(), access.num_elements(), final_data);
+    };
   }
 
+
+  /** Disable write-back on buffer destruction as an iterator.
+  */
+  void set_final_data(std::nullptr_t) {
+    shared_pointer = boost::none;
+    final_write_back = boost::none;
+  }
+
+
+  /** Provide destination for write-back on buffer destruction as an iterator.
+  */
+  template <typename Iterator>
+  void set_final_data(Iterator final_data) {
+    using type_ = typename iterator_value_type<Iterator>::value_type;
+    static_assert(std::is_same<type_, T>::value, "buffer type mismatch");
```

Comment:
Probably these tests might not work for some weird output iterators.
http://en.cppreference.com/w/cpp/iterator/back_insert_iterator